rlclientlib/federation/sender_joined_log_joiner.h

```diff
@@ -0,0 +1,248 @@
+#include "logger/preamble.h"
+#include "rl_string_view.h"
+#include "sender.h"
+#include "time_helper.h"
+#include "vw/common/text_utils.h"
+#include "vw/io/io_adapter.h"
+
+#include <flatbuffers/buffer.h>
+#include <flatbuffers/detached_buffer.h>
+#include <flatbuffers/flatbuffer_builder.h>
+
+#include <chrono>
+#include <cstdint>
+#include <map>
+#include <memory>
+#include <ratio>
+#include <string>
+#include <vector>
+
+namespace reinforcement_learning
+{
+
+  class buffer_reader : public VW::io::reader
+{
+public:
+  buffer_reader(std::vector<uint8_t>&& buffer)
+  : VW::io::reader(true),
+  _buffer(std::move(buffer)), _read_head(buffer.data())
+  {
+
+  }
+  ~buffer_reader() override = default;
+  ssize_t read(char* buffer, size_t num_bytes) override
+  {
+    num_bytes = std::min((_buffer.data() + _buffer.size()) - _read_head, static_cast<std::ptrdiff_t>(num_bytes));
+    if (num_bytes == 0) { return 0; }
+
+    std::memcpy(buffer, _read_head, num_bytes);
+    _read_head += num_bytes;
+
+    return num_bytes;
+  }
+  void reset() override
+  {
+    _read_head= _buffer.data();
+  }
+
+private:
+  std::vector<uint8_t> _buffer;
+  uint8_t* _read_head;
+};
+
+
+timestamp to_rl_timestamp(const reinforcement_learning::messages::flatbuff::v2::TimeStamp& ts)
+{
+  timestamp output;
+  output.year = ts.year();
+  output.day = ts.day();
+  output.month = ts.month();
+  output.hour = ts.hour();
+  output.minute = ts.minute();
+  output.second = ts.second();
+  output.sub_second = ts.subsecond();
+  return output;
+}
+
+reinforcement_learning::messages::flatbuff::v2::TimeStamp from_rl_timestamp(const timestamp& ts)
+{
+    return reinforcement_learning::messages::flatbuff::v2::TimeStamp(ts.year, ts.month, ts.day, ts.hour, ts.minute, ts.second, ts.sub_second);
+}
+
+int emit_filemagic_message(std::vector<uint8_t>& output)
+{
+  // TODO consider doing this a safer way
+  uint32_t filemagic = 0x42465756;
+  output.reserve(output.size() + 4);
+  output.insert(std::end(output), reinterpret_cast<uint8_t*>(&filemagic), reinterpret_cast<uint8_t*>(&filemagic)+4);
+  return 0;
+}
+
+int emit_regular_message(std::vector<uint8_t>& output, flatbuffers::FlatBufferBuilder& fbb, flatbuffers::Offset<reinforcement_learning::messages::flatbuff::v2::JoinedPayload> joined_payload)
+{
+  // TODO consider doing this a safer way
+  uint32_t regular = 0xFFFFFFFF;
+  output.reserve(output.size() + 4);
+  output.insert(std::end(output), reinterpret_cast<uint8_t*>(&regular), reinterpret_cast<uint8_t*>(&regular)+4);
+
+  fbb.Finish(joined_payload);
+  auto buffer = fbb.Release();
+
+  uint32_t size = buffer.size();
+  output.reserve(output.size() + 4);
+  output.insert(std::end(output), reinterpret_cast<uint8_t*>(&size), reinterpret_cast<uint8_t*>(&size)+4);
+
+  output.reserve(output.size() + size);
+  output.insert(std::end(output), buffer.data(), buffer.data()+size);
+  return 0;
+}
+
+struct sender_joined_log_provider : public i_joined_log_provider, i_sender
```

Comment:
does `sender_joined_log_provider` need to inherit `i_sender`? because we're only sending to it through a proxy object that contains a pointer to the joined_log_provider.

Reply:
Yeah it doesn't need it. Will remove